src/PkgAuthentication.jl

```diff
@@ -175,8 +175,31 @@ end
 Base.show(io::IO, s::NoAuthentication) = print(io, "NoAuthentication($(s.server))")
 
 function step(state::NoAuthentication)::Union{RequestLogin, Failure}
-    challenge = Random.randstring(32)
     output = IOBuffer()
+    # Try device flow first
+    response = Downloads.request(
+        joinpath(state.server, "dex/device/code"),
+        method = "POST",
+        input = IOBuffer("client_id=device&scope=openid email profile offline_access"),
+        output = output,
+        throw = false,
+        headers = Dict("Accept" => "application/json", "Content-Type" => "application/x-www-form-urlencoded"),
+    )
+    if response isa Downloads.Response && response.status == 200
+        body = nothing
+        try
+            body = JSON.parse(String(take!(output)))
+        catch ex
+            @debug "Request for device code returned: ", body
+        end
+
+        if body !== nothing
+            body["client_id"] = "device"
+            return RequestLogin(state.server, "", body)
+        end
+    end
+
+    challenge = Random.randstring(32)
```

Comment:
Refactor this block and the above device flow block out into two separate internal methods for clarity?

Reply:
Done in https://github.com/JuliaComputing/PkgAuthentication.jl/pull/42/commits/79e417d17ba77db3945a743853aa08d7a126385e